File: docs/lib/project-generator.ts
import JSZip from "jszip";
import type { ProjectOptions } from "./types";
import {
  generateBuildGradlePreview,
  generateSettingsGradlePreview,
  generateMainFilePreview,
  generateVersionCatalogPreview,
  generateReadmePreview,
} from "./preview-functions";

async function fetchTextFile(path: string): Promise<string> {
  try {
    const response = await fetch(
      `https://raw.githubusercontent.com/zahid4kh/compose-for-desktop/main/${path}`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch ${path}: ${response.status}`);
    }
    return await response.text();
  } catch (error) {
    console.error(`Error fetching ${path}:`, error);
    throw error;
  }
}

async function fetchBinaryFile(path: string): Promise<ArrayBuffer> {
  try {
    const response = await fetch(
      `https://raw.githubusercontent.com/zahid4kh/compose-for-desktop/main/${path}`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch ${path}: ${response.status}`);
    }
    return await response.arrayBuffer();
  } catch (error) {
    console.error(`Error fetching binary file ${path}:`, error);
    throw error;
  }
}

async function addGradleWrapperFiles(folder: JSZip): Promise<boolean> {
  try {
    // gradle-wrapper.properties
    const propertiesContent = await fetchTextFile(
      "gradle/wrapper/gradle-wrapper.properties"
    );
    folder.file("gradle-wrapper.properties", propertiesContent);

    // gradle-wrapper.jar (binary file)
    const jarContent = await fetchBinaryFile(
      "gradle/wrapper/gradle-wrapper.jar"
    );
    folder.file("gradle-wrapper.jar", jarContent);

    return true;
  } catch (error) {
    console.error("Error adding Gradle wrapper files:", error);
    throw error;
  }
}

async function addGradleWrapperScripts(folder: JSZip): Promise<boolean> {
  try {
    // gradlew (shell script)
    const gradlewContent = await fetchTextFile("gradlew");
    folder.file("gradlew", gradlewContent);

    // gradlew.bat (batch script)
    const gradlewBatContent = await fetchTextFile("gradlew.bat");
    folder.file("gradlew.bat", gradlewBatContent);

    return true;
  } catch (error) {
    console.error("Error adding Gradle wrapper scripts:", error);
    throw error;
  }
}

async function addGradleProperties(folder: JSZip): Promise<boolean> {
  try {
    const content = await fetchTextFile("gradle.properties");
    folder.file("gradle.properties", content);
    return true;
  } catch (error) {
    console.error("Error adding gradle.properties:", error);
    throw error;
  }
}

async function addKotlinFiles(folder: JSZip): Promise<boolean> {
  try {
    // App.kt
    const appContent = await fetchTextFile("src/main/kotlin/App.kt");
    folder.file("App.kt", appContent);

    // AppModule.kt
    const appModuleContent = await fetchTextFile(
      "src/main/kotlin/AppModule.kt"
    );
    folder.file("AppModule.kt", appModuleContent);

    // MainViewModel.kt
    const viewModelContent = await fetchTextFile(
      "src/main/kotlin/MainViewModel.kt"
    );
    folder.file("MainViewModel.kt", viewModelContent);

    // Models.kt
    const modelsContent = await fetchTextFile("src/main/kotlin/Models.kt");
    folder.file("Models.kt", modelsContent);

    return true;
  } catch (error) {
    console.error("Error adding Kotlin files:", error);
    throw error;
  }
}

async function addThemeFiles(folder: JSZip): Promise<boolean> {
  try {
    // Color.kt
    const colorContent = await fetchTextFile("src/main/kotlin/theme/Color.kt");
    folder.file("Color.kt", colorContent);

    // Theme.kt
    const themeContent = await fetchTextFile("src/main/kotlin/theme/Theme.kt");
    folder.file("Theme.kt", themeContent);

    // Type.kt
    const typeContent = await fetchTextFile("src/main/kotlin/theme/Type.kt");
    folder.file("Type.kt", typeContent);

    return true;
  } catch (error) {
    console.error("Error adding theme files:", error);
    throw error;
  }
}

async function addIconFiles(folder: JSZip): Promise<boolean> {
  try {
    // compose.png
    const pngContent = await fetchBinaryFile("icons/compose.png");
    folder.file("compose.png", pngContent);

    // compose.ico
    const icoContent = await fetchBinaryFile("icons/compose.ico");
    folder.file("compose.ico", icoContent);

    // compose.icns
    const icnsContent = await fetchBinaryFile("icons/compose.icns");
    folder.file("compose.icns", icnsContent);

    return true;
  } catch (error) {
    console.error("Error adding icon files:", error);
    throw error;
  }
}

async function addGitignoreFile(folder: JSZip): Promise<boolean> {
  try {
    const content = await fetchTextFile(".gitignore");
    folder.file(".gitignore", content);
    return true;
  } catch (error) {
    console.error("Error adding .gitignore file:", error);
    throw error;
  }
}

async function addDatabaseFile(
  folder: JSZip,
  options: { appName: string }
): Promise<void> {
  const content = `import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File

class Database {
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
    }

    private val appDir: File
    private val settingsFile: File

    init {
        val userHome = System.getProperty("user.home")
        appDir = File(userHome, ".${options.appName
          .toLowerCase()
          .replace(/\s+/g, "")}").apply {
            if (!exists()) mkdirs()
        }

        settingsFile = File(appDir, "settings.json")

        if (!settingsFile.exists()) settingsFile.writeText(json.encodeToString(AppSettings()))
    }

    suspend fun getSettings(): AppSettings = withContext(Dispatchers.IO) {
        return@withContext try {
            json.decodeFromString(settingsFile.readText())
        } catch (e: Exception) {
            AppSettings()
        }
    }

    suspend fun saveSettings(settings: AppSettings) = withContext(Dispatchers.IO) {
        settingsFile.writeText(json.encodeToString(settings))
    }
}`;

  folder.file("Database.kt", content);
}

export async function generateProject(options: ProjectOptions) {
  try {
    console.log("Starting project generation...", options);

    const zip = new JSZip();

    const appNameFormatted = options.appName.toLowerCase().replace(/\s+/g, "-");
    console.log("Formatted app name:", appNameFormatted);

    const rootFolder = zip.folder(appNameFormatted);
    if (!rootFolder) {
      throw new Error("Failed to create root folder");
    }

    console.log("Creating project structure...");

    // folder structure
    const gradleFolder = rootFolder.folder("gradle");
    if (!gradleFolder) {
      throw new Error("Failed to create gradle folder");
    }

    const wrapperFolder = gradleFolder.folder("wrapper");
    if (!wrapperFolder) {
      throw new Error("Failed to create wrapper folder");
    }

    const iconsFolder = rootFolder.folder("icons");
    if (!iconsFolder) {
      throw new Error("Failed to create icons folder");
    }

    const srcFolder = rootFolder.folder("src");
    if (!srcFolder) {
      throw new Error("Failed to create src folder");
    }

    const mainFolder = srcFolder.folder("main");
    if (!mainFolder) {
      throw new Error("Failed to create main folder");
    }

    const composeResourcesFolder = mainFolder.folder("composeResources");
    if (!composeResourcesFolder) {
      throw new Error("Failed to create composeResources folder");
    }

    const drawableFolder = composeResourcesFolder.folder("drawable");
    if (!drawableFolder) {
      throw new Error("Failed to create drawable folder");
    }

    const kotlinFolder = mainFolder.folder("kotlin");
    if (!kotlinFolder) {
      throw new Error("Failed to create kotlin folder");
    }

    const themeFolder = kotlinFolder.folder("theme");
    if (!themeFolder) {
      throw new Error("Failed to create theme folder");
    }

    mainFolder.folder("resources");

    const testFolder = srcFolder.folder("test");
    if (testFolder) {
      testFolder.folder("kotlin");
      testFolder.folder("resources");
    }

    console.log("Adding Gradle files...");
    // Add Gradle files
    await addGradleWrapperFiles(wrapperFolder);
    await addGradleWrapperScripts(rootFolder);
    await addGradleProperties(rootFolder);

    // Add version catalog
    const versionCatalog = generateVersionCatalogPreview(options);
    gradleFolder.file("libs.versions.toml", versionCatalog);
    console.log("Version catalog generated");

    // Add build.gradle.kts
    const buildGradle = generateBuildGradlePreview(options);
    rootFolder.file("build.gradle.kts", buildGradle);
    console.log("Build gradle generated");

    // Add settings.gradle.kts
    const settingsGradle = generateSettingsGradlePreview(options);
    rootFolder.file("settings.gradle.kts", settingsGradle);
    console.log("Settings gradle generated");

    console.log("Adding Icon files...");
    // Add icon files
    await addIconFiles(iconsFolder);

    console.log("Adding Kotlin files...");
    // Add Main.kt
    const mainKt = generateMainFilePreview(options);
    kotlinFolder.file("Main.kt", mainKt);
    console.log("Main.kt generated");

    // Add Database.kt
    await addDatabaseFile(kotlinFolder, { appName: options.appName });
    console.log("Database.kt generated");

    // Add other Kotlin files
    await addKotlinFiles(kotlinFolder);
    console.log("Additional Kotlin files generated");

    // Add theme files
    await addThemeFiles(themeFolder);
    console.log("Theme files generated");

    // Add README.md
    const readme = generateReadmePreview(options);
    rootFolder.file("README.md", readme);
    console.log("README.md generated");

    // Add .gitignore
    await addGitignoreFile(rootFolder);
    console.log(".gitignore generated");

    console.log("Generating ZIP file...");

    const content = await zip.generateAsync({
      type: "blob",
      compression: "DEFLATE",
      compressionOptions: { level: 9 },
    });
    console.log("ZIP file generated, size:", content.size);

    console.log("Creating download link...");
    const url = URL.createObjectURL(content);
    const link = document.createElement("a");
    link.href = url;
    link.download = `${appNameFormatted}.zip`;

    document.body.appendChild(link);
    console.log("Triggering download...");

    setTimeout(() => {
      link.click();

      console.log("Cleaning up...");
      setTimeout(() => {
        URL.revokeObjectURL(url);
        document.body.removeChild(link);
        console.log("Download process completed");
      }, 200);
    }, 100);

    return true;
  } catch (error) {
    console.error("Error generating project:", error);
    alert(
      `Error generating project: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
    throw error;
  }
}

File: desktopclient/src/main/kotlin/ProjectGenerator.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.nio.file.Files
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import javax.swing.JFileChooser
import javax.swing.filechooser.FileNameExtensionFilter

class ProjectGenerator {

    suspend fun generateProject(options: ProjectOptions): String = withContext(Dispatchers.IO) {
        val appNameFormatted = options.appName.lowercase().replace(Regex("\\s+"), "-")
        val destinationPath = chooseDestination(appNameFormatted)
        val tempDir = Files.createTempDirectory("compose-desktop-").toFile()

        try {
            val rootDir = File(tempDir, appNameFormatted)
            rootDir.mkdirs()

            // dirs
            val gradleDir = File(rootDir, "gradle")
            val wrapperDir = File(gradleDir, "wrapper")
            val iconsDir = File(rootDir, "icons")
            val srcDir = File(rootDir, "src")
            val mainDir = File(srcDir, "main")
            val kotlinDir = File(mainDir, "kotlin")
            val themeDir = File(kotlinDir, "theme")
            val resourcesDir = File(mainDir, "resources")
            val composeResourcesDir = File(mainDir, "composeResources")
            val drawableDir = File(composeResourcesDir, "drawable")

            listOf(gradleDir, wrapperDir, iconsDir, srcDir, mainDir, kotlinDir,
                themeDir, resourcesDir, composeResourcesDir, drawableDir).forEach { it.mkdirs() }

            // Generate and write text files
            writeTextFile(File(rootDir, "build.gradle.kts"), PreviewFunctions.generateBuildGradlePreview(options))
            writeTextFile(File(rootDir, "settings.gradle.kts"), PreviewFunctions.generateSettingsGradlePreview(options))
            writeTextFile(File(gradleDir, "libs.versions.toml"), PreviewFunctions.generateVersionCatalogPreview(options))
            writeTextFile(File(kotlinDir, "Main.kt"), PreviewFunctions.generateMainFilePreview(options))
            writeTextFile(File(rootDir, "README.md"), PreviewFunctions.generateReadmePreview(options))

            // Fetch and write static files
            fetchAndWriteTextFile("gradle/wrapper/gradle-wrapper.properties", File(wrapperDir, "gradle-wrapper.properties"))
            fetchAndWriteBinaryFile("gradle/wrapper/gradle-wrapper.jar", File(wrapperDir, "gradle-wrapper.jar"))
            fetchAndWriteTextFile("gradlew", File(rootDir, "gradlew"))
            fetchAndWriteTextFile("gradlew.bat", File(rootDir, "gradlew.bat"))
            fetchAndWriteTextFile("gradle.properties", File(rootDir, "gradle.properties"))
            fetchAndWriteTextFile(".gitignore", File(rootDir, ".gitignore"))

            // Kotlin files
            fetchAndWriteTextFile("src/main/kotlin/App.kt", File(kotlinDir, "App.kt"))
            fetchAndWriteTextFile("src/main/kotlin/AppModule.kt", File(kotlinDir, "AppModule.kt"))
            fetchAndWriteTextFile("src/main/kotlin/MainViewModel.kt", File(kotlinDir, "MainViewModel.kt"))
            fetchAndWriteTextFile("src/main/kotlin/Models.kt", File(kotlinDir, "Models.kt"))

            // Theme files
            fetchAndWriteTextFile("src/main/kotlin/theme/Color.kt", File(themeDir, "Color.kt"))
            fetchAndWriteTextFile("src/main/kotlin/theme/Theme.kt", File(themeDir, "Theme.kt"))
            fetchAndWriteTextFile("src/main/kotlin/theme/Type.kt", File(themeDir, "Type.kt"))

            // Icon files
            fetchAndWriteBinaryFile("icons/compose.png", File(iconsDir, "compose.png"))
            fetchAndWriteBinaryFile("icons/compose.ico", File(iconsDir, "compose.ico"))
            fetchAndWriteBinaryFile("icons/compose.icns", File(iconsDir, "compose.icns"))

            // Database.kt with template
            writeDatabaseFile(kotlinDir, options)

            // Proguard rules
            writeProguardRules(rootDir)

            // Create ZIP file
            createZipFile(rootDir, destinationPath)

            return@withContext destinationPath.absolutePath

        } finally {
            tempDir.deleteRecursively()
        }
    }

    private fun chooseDestination(defaultName: String): File {
        val fileChooser = JFileChooser().apply {
            fileSelectionMode = JFileChooser.FILES_ONLY
            val downloadsDir = File(System.getProperty("user.home"), "Downloads")
            selectedFile = File(downloadsDir, "$defaultName.zip")
            fileFilter = FileNameExtensionFilter("ZIP files", "zip")
        }

        return if (fileChooser.showSaveDialog(null) == JFileChooser.APPROVE_OPTION) {
            val file = fileChooser.selectedFile
            if (file.extension != "zip") {
                File(file.absolutePath + ".zip")
            } else {
                file
            }
        } else {
            val downloadsDir = File(System.getProperty("user.home"), "Downloads")
            File(downloadsDir, "$defaultName.zip")
        }
    }

    private fun writeTextFile(file: File, content: String) {
        file.writeText(content)
    }

    private suspend fun fetchAndWriteTextFile(path: String, destination: File) = withContext(Dispatchers.IO) {
        val url = URL("https://raw.githubusercontent.com/zahid4kh/compose-for-desktop/main/$path")
        url.readText().let { destination.writeText(it) }
        if (path.endsWith("gradlew") && !System.getProperty("os.name").lowercase().contains("windows")) {
            try {
                destination.setExecutable(true)
            } catch (e: Exception) {
                println("Warning: Failed to set executable permission on ${destination.absolutePath}")
                println(e.message)
            }
        }

    }

    private suspend fun fetchAndWriteBinaryFile(path: String, destination: File) = withContext(Dispatchers.IO) {
        val url = URL("https://raw.githubusercontent.com/zahid4kh/compose-for-desktop/main/$path")
        url.openStream().use { input ->
            destination.outputStream().use { output ->
                input.copyTo(output)
            }
        }
    }

    private fun writeDatabaseFile(kotlinDir: File, options: ProjectOptions) {
        val content = """import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File

class Database {
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
    }

    private val appDir: File
    private val settingsFile: File

    init {
        val userHome = System.getProperty("user.home")
        appDir = File(userHome, ".${options.appName.lowercase().replace(Regex("\\s+"), "")}").apply {
            if (!exists()) mkdirs()
        }

        settingsFile = File(appDir, "settings.json")

        if (!settingsFile.exists()) settingsFile.writeText(json.encodeToString(AppSettings()))
    }

    suspend fun getSettings(): AppSettings = withContext(Dispatchers.IO) {
        return@withContext try {
            json.decodeFromString(settingsFile.readText())
        } catch (e: Exception) {
            AppSettings()
        }
    }

    suspend fun saveSettings(settings: AppSettings) = withContext(Dispatchers.IO) {
        settingsFile.writeText(json.encodeToString(settings))
    }
}"""

        File(kotlinDir, "Database.kt").writeText(content)
    }

    private fun writeProguardRules(rootDir: File) {
        val content = """-dontwarn kotlinx.serialization.**

-dontwarn sun.font.CFont
-dontwarn sun.swing.SwingUtilities2${'$'}AATextInfo
-dontwarn net.miginfocom.swing.MigLayout

-dontnote kotlinx.serialization.**
-dontnote META-INF.**
-dontnote kotlinx.serialization.internal.PlatformKt

# Keep Kotlin Serialization
-keepattributes *Annotation*, InnerClasses
-dontnote kotlinx.serialization.AnnotationsKt

# Keep all serializable classes with their @Serializable annotation
-keepclassmembers class ** {
    @kotlinx.serialization.Serializable <fields>;
}

# Keep serializers
-keepclasseswithmembers class **${'$'}${'$'}serializer {
    static **${'$'}${'$'}serializer INSTANCE;
}


# Keep serializable classes and their properties
-if @kotlinx.serialization.Serializable class **
-keep class <1> {
    static <1>${'$'}Companion Companion;
}

# Keep specific serializer classes
-keepclassmembers class kotlinx.serialization.json.** {
    *** Companion;
}
-keepclasseswithmembers class kotlinx.serialization.json.** {
    kotlinx.serialization.KSerializer serializer(...);
}

# Keep serialization descriptors
-keep class kotlinx.serialization.descriptors.** { *; }

# Specifically keep AppSettings and its serializer
-keep class AppSettings { *; }
-keep class AppSettings${'$'}${'$'}serializer { *; }"""

        File(rootDir, "proguard-rules.pro").writeText(content)
    }

    private fun createZipFile(sourceDir: File, destination: File) {
        ZipOutputStream(BufferedOutputStream(FileOutputStream(destination))).use { zipOut ->
            sourceDir.walkTopDown().forEach { file ->
                val relativePath = file.relativeTo(sourceDir).path
                
                if (file == sourceDir) {
                    return@forEach
                }
                
                val entryPath = relativePath.replace("\\", "/")
                
                if (file.isDirectory) {
                    val directoryEntryPath = if (entryPath.endsWith("/")) entryPath else "$entryPath/"
                    zipOut.putNextEntry(ZipEntry(directoryEntryPath))
                } else {
                    zipOut.putNextEntry(ZipEntry(entryPath))
                    file.inputStream().use { it.copyTo(zipOut) }
                }
            }
        }
    }
}